Parser crashes from fuzzing

[bork1.txt](https://github.com/chitin-io/chitin/files/100114/bork1.txt)

```
"chitin\tv1message 0\tv0{slots{}"
```

```
panic: interface conversion: interface is nil, not parser.MatchTree

goroutine 1 [running]:
chitin.io/chitin/schema.makeParser.func9(0x8349a0, 0xc8200a55a0, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/chitin.io/chitin/schema/parser.go:151 +0x402
github.com/andyleap/parser.(*Grammar).Node.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:42 +0x108
github.com/andyleap/parser.Tag.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:203 +0xb0
github.com/andyleap/parser.Require.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:187 +0x195
github.com/andyleap/parser.And.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:109 +0x195
github.com/andyleap/parser.Tag.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:203 +0xb0
github.com/andyleap/parser.And.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:109 +0x195
github.com/andyleap/parser.Or.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:128 +0x198
github.com/andyleap/parser.Mult.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:151 +0x16b
github.com/andyleap/parser.And.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:109 +0x195
github.com/andyleap/parser.(*Grammar).Node.func1(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:40 +0x90
github.com/andyleap/parser.(*Grammar).Parse(0xc820038ec8, 0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:52 +0x83
chitin.io/chitin/schema.Parse(0x7fdc32bed268, 0xc8200dbe30, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/chitin.io/chitin/schema/parser.go:13 +0x61
chitin.io/chitin/schema.Fuzz(0x7fdc311aa000, 0x1d, 0x200000, 0x3)
	/home/tv/tmp/go-fuzz-build999057326/src/chitin.io/chitin/schema/fuzz.go:21 +0xe9
github.com/dvyukov/go-fuzz/go-fuzz-dep.Main(0x9f2790)
	/home/tv/go/src/github.com/dvyukov/go-fuzz/go-fuzz-dep/main.go:47 +0x14c
main.main()
	/home/tv/tmp/go-fuzz-build999057326/src/go-fuzz-main/main.go:10 +0x23

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/home/tv/tmp/go-fuzz-build999057326/src/runtime/asm_amd64.s:1721 +0x1
```

---

[bork2.txt](https://github.com/chitin-io/chitin/files/100115/bork2.txt)

```
"chitin\tv1envelope\t0{}"
```

```
panic: interface conversion: parser.Match is nil, not schema.Envelope

goroutine 1 [running]:
chitin.io/chitin/schema.makeParser.func11(0x8349a0, 0xc8200c0c40, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/chitin.io/chitin/schema/parser.go:222 +0x5c7
github.com/andyleap/parser.(*Grammar).Node.func1(0x7f612b677700, 0xc8205141b0, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:42 +0x108
github.com/andyleap/parser.(*Grammar).Parse(0xc8200398d8, 0x7f612b677700, 0xc8205141b0, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/github.com/andyleap/parser/parser.go:52 +0x83
chitin.io/chitin/schema.Parse(0x7f612b677700, 0xc8205141b0, 0x0, 0x0, 0x0, 0x0)
	/home/tv/tmp/go-fuzz-build999057326/src/chitin.io/chitin/schema/parser.go:13 +0x61
chitin.io/chitin/schema.Fuzz(0x7f6129c70000, 0x15, 0x200000, 0x3)
	/home/tv/tmp/go-fuzz-build999057326/src/chitin.io/chitin/schema/fuzz.go:21 +0xe9
github.com/dvyukov/go-fuzz/go-fuzz-dep.Main(0x9f2790)
	/home/tv/go/src/github.com/dvyukov/go-fuzz/go-fuzz-dep/main.go:47 +0x14c
main.main()
	/home/tv/tmp/go-fuzz-build999057326/src/go-fuzz-main/main.go:10 +0x23

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/home/tv/tmp/go-fuzz-build999057326/src/runtime/asm_amd64.s:1721 +0x1
```

--- a/schema/parser.go
+++ b/schema/parser.go
@@ -92,17 +92,21 @@ func makeParser() *p.Grammar {
 	slotItemFormat := p.And(
 		p.Tag("name", p.Mult(1, 0, p.Set("\\w"))), p.Require(ws, p.Tag("type", types), ws))
 	slotItemFormat.Node(func(m p.Match) (p.Match, error) {
+		slottype, ok := p.GetTag(m, "type").(SlotType)
+		if !ok {
+			return nil, fmt.Errorf("%T is not a SlotType", p.GetTag(m, "type"))
+		}
 		slot := Slot{
 			Name: p.String(p.GetTag(m, "name")),
-			Kind: p.GetTag(m, "type").(SlotType),
+			Kind: slottype,
 		}
-		return slot, nil
+		return p.TagMatch("slot", slot), nil
 	})
 
 	slotFormat := p.And(
 		p.Lit("slots"), p.Require(ws,
 			p.Lit("{"), ws,
-			p.Tag("slots", p.Mult(0, 0, p.And(slotItemFormat))),
+			p.Mult(0, 0, p.And(slotItemFormat)),
 			p.Lit("}"),
 		))
 
@@ -113,13 +117,13 @@ func makeParser() *p.Grammar {
 			Name: p.String(p.GetTag(m, "name")),
 			Kind: p.GetTag(m, "type").(FieldType),
 		}
-		return field, nil
+		return p.TagMatch("field", field), nil
 	})
 
 	fieldFormat := p.And(
 		p.Lit("fields"), p.Require(ws,
 			p.Lit("{"), ws,
-			p.Tag("fields", p.Mult(0, 0, p.And(fieldItemFormat))),
+			p.Mult(0, 0, p.And(fieldItemFormat)),
 			p.Lit("}"),
 		))
 
@@ -144,12 +148,12 @@ func makeParser() *p.Grammar {
 		wireFormat, _ := p.GetTag(m, "wireFormat").(uint32)
 
 		slots := []Slot{}
-		for _, slot := range p.GetTag(m, "slots").(p.MatchTree) {
-			slots = append(slots, slot.(p.MatchTree)[0].(Slot))
+		for _, slot := range p.GetTags(m, "slot") {
+			slots = append(slots, slot.(Slot))
 		}
 		fields := []Field{}
-		for _, field := range p.GetTag(m, "fields").(p.MatchTree) {
-			fields = append(fields, field.(p.MatchTree)[0].(Field))
+		for _, field := range p.GetTags(m, "field") {
+			fields = append(fields, field.(Field))
 		}
 		if len(fields) == 0 {
 			msg := &FixedMessage{
@@ -219,7 +223,12 @@ func makeParser() *p.Grammar {
 		envelopes := make(map[string]Envelope)
 		for _, msg := range p.GetTags(m, "envelope") {
 			name := p.String(p.GetTag(msg, "name"))
-			envelopes[name] = p.GetTag(msg, "body").(Envelope)
+			var e Envelope
+			ok := false
+			if e, ok = p.GetTag(msg, "body").(Envelope); !ok {
+				return nil, fmt.Errorf("Envelope %s has no mapping!", name)
+			}
+			envelopes[name] = e
 		}
 
 		s := &SchemaV1{

--- a/schema/parser_test.go
+++ b/schema/parser_test.go
@@ -1,0 +1,23 @@
+package schema
+
+import "testing"
+
+func TestFuzz1(t *testing.T) {
+	parser := makeParser()
+
+	parser.ParseString("chitin\tv1message 0\tv0{slots{}")
+
+}
+
+func TestFuzz2(t *testing.T) {
+	parser := makeParser()
+
+	parser.ParseString("chitin\tv1envelope\t0{}")
+}
+
+func TestFuzz3(t *testing.T) {
+	parser := makeParser()
+
+	parser.ParseString("chitin\nv1message\n0\nv0{slots{0\nstring")
+
+}

